content/post/101-rust-ipfs-unixfs.md

```diff
@@ -0,0 +1,152 @@
+can be a long time before any of the previously seen "prefetchable" blocks are
+ready to be processed.
+
+Continuing the walk returns a `ContinuedWalk` on success, which is an enum of
+three possibilities:
+
+1. file (new or continuation)
+2. directory (new or continuation)
+3. symlink
+
+For files and symlinks, there is an additional [`FileSegment`] or a slice of bytes
+provided. This represents a segment of the file, which could possibly be empty.
+For symlinks the slice is the linked path of the `Symlink`.
+
+From a `ContinuedWalk` value the walk can be continued by first accessing the
+inner `Item` value either by pattern matching or by using
+`ContinuedWalk::into_inner`. The [`Item::into_inner`] will return an `Option`
+of the next `Walker`. One will exist if there are still links to walk.
+
+[MerkleDAG is the outer protocol buffers description]: https://github.com/ipfs/go-merkledag/blob/master/pb/merkledag.proto
+[UnixFS messages]: https://github.com/ipfs/specs/blob/master/UNIXFS.md
+[`FileSegment`]: https://docs.rs/ipfs-unixfs/0.0.1/ipfs_unixfs/walk/struct.FileSegment.html
+[issue #200]: https://github.com/rs-ipfs/rust-ipfs/issues/200
+[`Item::into_inner`]: https://docs.rs/ipfs-unixfs/0.0.1/ipfs_unixfs/walk/struct.Item.html#method.into_inner
+
+## `/cat` and `/get`
+
+In short, these two operations **retrieve** data from MerkleDAGs and, in the
+process, fetch the required blocks from the network. [`/cat`] can only process
+UnixFS blocks of type `File` or `Raw`, while [`/get`] can start from a
+block of type `Directory`, `File` or `Symlink`.
+
+Depending on the type of the block, the walk will only consist of a single
+`File` or `Symlink` block, a single `File` tree, or some combination,
+including sub-directories in a directory hierarchy. The `/cat` HTTP API returns
+the raw file whereas `/get` ([HTTP impl](https://github.com/rs-ipfs/rust-ipfs/blob/a1614011a330d32842352bf1095219e6b068b92a/http/src/v0/root_files.rs#L86-L195))
+returns a tar file of the tree.
+
+Exposing the `/cat` operation ([HTTP impl](https://github.com/rs-ipfs/rust-ipfs/blob/a1614011a330d32842352bf1095219e6b068b92a/http/src/v0/root_files.rs#L38-L69))
+as a Rust API was the easier part, as it is a pretty good fit for a
+`futures::stream::Stream<Item = Result<Vec<u8>, _>>` or
+`futures::stream::TryStream<Ok = Vec<u8>, Error = _>`. As we had already used
+`async_stream` over at `ipfs-http` for the initial `refs` implementation, it
+again seemed like the easiest way to get this implemented.
+
+The implementation ended up using `Vec<u8>` for bytes instead of `bytes::Bytes`
+simply because we currently didn't depend on `bytes` even indirectly in `ipfs`.
+It might be that the type in the signature should be changed to
+`impl AsRef<[u8]>` to allow later migration to `Bytes`.
+
+[`/cat`]: https://docs.ipfs.io/reference/http/api/#api-v0-cat
+[`/get`]: https://docs.ipfs.io/reference/http/api/#api-v0-get
+
+File metadata could potentially be handled by first returning a
```

Comment:
instead of ending with potential ways to change your implmementation - this is a great place to explain what people can do with rust-ipfs at its current level of functionality, and what you expect to be the first projects building on rust-ipfs

Reply:
Good call, will do. :+1: